Firestore/core/src/firebase/firestore/model/blob.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/blob.h"
+
+#include <string.h>
+
+#include <algorithm>
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+Blob::Blob() : buffer_(nullptr), size_(0) {
+}
+
+Blob::Blob(const Blob& value) : Blob() {
+  *this = value;
+}
+
+Blob::Blob(Blob&& value) : Blob() {
+  *this = std::move(value);
+}
+
+Blob::~Blob() {
+  delete[] buffer_;
+}
+
+Blob Blob::CopyFrom(const uint8_t* source, size_t size) {
+  uint8_t* copy = new uint8_t[size];
```

Comment:
Since we're always copying now, is there any reason not to make our internal buffer just use `std::vector<uint8_t>`?

At that point we stop having to worry about managing pointers at all.

Given an external source and size, you can construct the vector directly:

```c++
std::vector<uint8_t> buffer(source, source + size);
```

Reply:
What you're saying is true (that we can't move from something else while using vector) but the reality is that the the public API is going to be `firebase::Variant` which has two kinds of values:

  * "mutable" blobs, whose values were allocated by the Variant
  * "static" blobs, whose values were passed by the caller and the caller still owns

You can't move from either of these:

  * In the "mutable" case the user has declared the variant to have a type of value and we can't change that type. Therefore the buffer is owned by the Variant.
  * In the "static" case we can't take control of the user buffer.

Reply:
This seems like a reasonable conclusion.